docstore/query.go

```diff
@@ -196,20 +197,100 @@ func (q *Query) Delete(ctx context.Context) error {
 	if err := q.validateWrite("delete"); err != nil {
 		return err
 	}
-	return q.coll.driver.RunDeleteQuery(ctx, q.dq)
+	type rdq interface {
+		RunDeleteQuery(context.Context, *driver.Query) error
+	}
+
+	if d, ok := q.coll.driver.(rdq); ok {
+		return d.RunDeleteQuery(ctx, q.dq)
+	}
+
+	return q.runAction(ctx, func(al *ActionList, doc Document) {
+		al.Delete(doc)
+	})
+}
+
+func (q *Query) runAction(ctx context.Context, addAction func(*ActionList, Document)) error {
+	var retries map[interface{}]bool
+	// TODO(shantuo): split actions into groups to reduce client memory.
+	// TODO(shantuo): test retry logic (using a mock driver)
+	for {
+		// Run the query. For each document it returns, add an action to an ActionList.
+		// TODO(shantuo): fetch only key and revision fields? Not possible
+		// if the collection uses a key-extraction function.
+		iter := q.Get(ctx)
+		al := q.coll.Actions()
+		var keys []interface{}
+		for {
+			doc := map[string]interface{}{}
+			err := iter.Next(ctx, doc)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			ddoc, err := driver.NewDocument(doc)
+			if err != nil {
+				return err
+			}
+			key, err := q.coll.driver.Key(ddoc)
+			if err != nil {
+				return err
+			}
+			// If we're retrying, only act on documents that need to be retried.
+			if retries == nil || retries[key] {
+				addAction(al, doc)
+				keys = append(keys, key)
+			}
+
+		}
+		// Run the Action List.
+		err := al.Do(ctx)
+		// If it succeeds, we're done.
+		if err == nil {
+			return nil
+		}
+		alerr := err.(ActionListError)
+		// If there are any errors that are not due to revision
+		// mismatches, fail immediately.
+		for _, e := range alerr {
+			if gcerrors.Code(e.Err) != gcerrors.FailedPrecondition {
+				return err
+			}
+		}
+		// Collect the keys of all documents that failed due to revision
+		// mismatches.
+		retries := map[interface{}]bool{}
+		for _, e := range alerr {
+			retries[keys[e.Index]] = true
+		}
+		// Re-run the query.
+	}
 }
 
 // Update updates all the documents specified by the query.
 // It is an error if the query has a limit.
+// TODO: export after fixing #2611.
```

Comment:
Should this be un-exported now?

Reply:
Left for a follow-up CL (by you :). The changes are extensive (rewriting the How-To, for example) and I don't have the bandwidth for them.